pkg/function/create_csi_snapshot.go

```diff
@@ -0,0 +1,138 @@
+	"github.com/kanisterio/kanister/pkg/kube"
+	"github.com/kanisterio/kanister/pkg/kube/snapshot"
+	"github.com/kanisterio/kanister/pkg/param"
+	"github.com/pkg/errors"
+)
+
+func init() {
+	_ = kanister.Register(&createCSISnapshotFunc{})
+}
+
+var (
+	_ kanister.Func = (*createCSISnapshotFunc)(nil)
+)
+
+const (
+	// CreateCSIVolumeSnapshotFuncName gives the name of the function
+	CreateCSISnapshotFuncName = "CreateCSISnapshot"
+	// CreateCSISnapshotNameArg provides name of the new VolumeSnapshot
+	CreateCSISnapshotNameArg = "name"
+	// CreateCSISnapshotPVCNameArg gives the name of the captured PVC
+	CreateCSISnapshotPVCNameArg = "pvc"
+	// CreateCSISnapshotNamespaceArg mentions the namespace of the captured PVC
+	CreateCSISnapshotNamespaceArg = "namespace"
+	// CreateCSISnapshotSnapshotClassArg specifies the name of the VolumeSnapshotClass
+	CreateCSISnapshotSnapshotClassArg = "snapshotClass"
+	// CreateCSISnapshotLabelsArg has labels that are to be added to the new VolumeSnapshot
+	CreateCSISnapshotLabelsArg = "labels"
+	// CreateCSISnapshotRestoreSizeArg gives the storage size required for PV/PVC restoration
+	CreateCSISnapshotRestoreSizeArg = "restoreSize"
+	// CreateCSISnapshotSnapshotContentNameArg provides the name of dynamically provisioned VolumeSnapshotContent
+	CreateCSISnapshotSnapshotContentNameArg = "snapshotContent"
+	// CreateCSISnapshotDefaultTimeout is the time duration in minutes for VolumeSnapshot to be ReadyToUse before context is timed out
+	CreateCSISnapshotDefaultTimeout = 2 * time.Minute
+)
+
+type createCSISnapshotFunc struct{}
+
+func (*createCSISnapshotFunc) Name() string {
+	return CreateCSISnapshotFuncName
+}
+
+func (*createCSISnapshotFunc) Exec(ctx context.Context, tp param.TemplateParams, args map[string]interface{}) (map[string]interface{}, error) {
+	var snapshotClass string
+	var labels map[string]string
+	var name, pvc, namespace string
+	if err := Arg(args, CreateCSISnapshotPVCNameArg, &pvc); err != nil {
+		return nil, err
+	}
+	if err := Arg(args, CreateCSISnapshotNamespaceArg, &namespace); err != nil {
+		return nil, err
+	}
+	if err := Arg(args, CreateCSISnapshotSnapshotClassArg, &snapshotClass); err != nil {
+		return nil, err
+	}
+	if err := OptArg(args, CreateCSISnapshotNameArg, &name, defaultSnapshotName(pvc, 20)); err != nil {
+		return nil, err
+	}
+	if err := OptArg(args, CreateCSISnapshotLabelsArg, &labels, map[string]string{}); err != nil {
+		return nil, err
+	}
+
+	kubeCli, err := kube.NewClient()
+	if err != nil {
+		return nil, err
+	}
+	dynCli, err := kube.NewDynamicClient()
+	if err != nil {
+		return nil, err
+	}
+	snapshotter, err := snapshot.NewSnapshotter(kubeCli, dynCli)
+	if err != nil {
+		return nil, err
+	}
+	// waitForReady is set to true by default because snapshot information is needed as output artifacts
+	waitForReady := true
+	Tomb := &tomb.Tomb{}
+	ctx, cancel := context.WithTimeout(Tomb.Context(ctx), CreateCSISnapshotDefaultTimeout)
+	defer cancel()
```

Comment:
Let's remove this. The timeout will be set by the action executor

Reply:
Done